src/transaction/log_recovery.c

```diff
@@ -572,18 +592,38 @@ template <typename T>
 void log_rv_redo_record_sync_or_dispatch_async (THREAD_ENTRY * thread_p, log_reader & log_pgptr_reader,
                                                 const T & log_rec, const log_lsa & rcv_lsa,
                                                 const LOG_LSA * end_redo_lsa, LOG_RECTYPE log_rtype,
-                                                LOG_ZIP & undo_unzip_support, LOG_ZIP & redo_unzip_support)
+                                                LOG_ZIP & undo_unzip_support, LOG_ZIP & redo_unzip_support,
+                                                cublog::ux_redo_parallel& parallel_recovery_redo)
 {
   const VPID rcv_vpid = log_rv_get_log_rec_vpid<T> (log_rec);
   // at this point, vpid can either be valid or not
 
   const LOG_DATA &log_data = log_rv_get_log_rec_data<T> (log_rec);
+  const bool need_sync_redo = log_rv_need_sync_redo (log_data.rcvindex);
+  assert(log_data.rcvindex != RVDK_UNRESERVE_SECTORS || need_sync_redo);
 
   // once vpid is extracted (or not), and depending on parameters, either dispatch the applying of
   // log redo asynchronously, or invoke synchronously
+  if (parallel_recovery_redo == nullptr || VPID_ISNULL (&rcv_vpid) || need_sync_redo)
+    {
+      // TODO: workaround, for few functions, make sure all dispatched actions have finished to prevent
+      // data race conditions
+      if (parallel_recovery_redo != nullptr && log_data.rcvindex == RVDK_UNRESERVE_SECTORS)
```

Comment:
@ribram 
I need a more semantically-strong explanation for this exception (see PR description).

Reply:
> To apply RVDK_UNRESERVE_SECTORS, one must first wait for all changes in this sector to be redone.
Otherwise, asynchronous jobs skip redoing changes in this sector's pages because they are seen as deallocated. When sector is reserved again, redo is resumed in the sector's pages, but the pages are not in a consistent state.
The current workaround is to wait for all changes to be finished, including changes in the unreserved sector.

Reply:
Thanks